src/goto-instrument/contracts/cfg_info.h

```diff
@@ -0,0 +1,203 @@
+  /// and does not contain `dereference` or `address_of` operations.
+  virtual bool is_local_composite_access(const exprt &expr) const final
+  {
+    // case-splitting on the lhs structure copied from symex_assignt::assign_rec
+    if(expr.id() == ID_symbol)
+    {
+      return is_local(to_symbol_expr(expr).get_identifier());
+    }
+    else if(expr.id() == ID_index)
+    {
+      return is_local_composite_access(to_index_expr(expr).array());
+    }
+    else if(expr.id() == ID_member)
+    {
+      const typet &type = to_member_expr(expr).struct_op().type();
+      if(
+        type.id() == ID_struct || type.id() == ID_struct_tag ||
+        type.id() == ID_union || type.id() == ID_union_tag)
+      {
+        return is_local_composite_access(to_member_expr(expr).compound());
+      }
+      else
+      {
+        throw unsupported_operation_exceptiont(
+          "is_local_composite_access: unexpected assignment to member of '" +
+          type.id_string() + "'");
+      }
+    }
+    else if(expr.id() == ID_if)
+    {
+      return is_local_composite_access(to_if_expr(expr).true_case()) &&
+             is_local_composite_access(to_if_expr(expr).false_case());
+    }
+    else if(expr.id() == ID_typecast)
+    {
+      return is_local_composite_access(to_typecast_expr(expr).op());
+    }
+    else if(
+      expr.id() == ID_byte_extract_little_endian ||
+      expr.id() == ID_byte_extract_big_endian)
+    {
+      return is_local_composite_access(to_byte_extract_expr(expr).op());
+    }
+    else if(expr.id() == ID_complex_real)
+    {
+      return is_local_composite_access(to_complex_real_expr(expr).op());
+    }
+    else if(expr.id() == ID_complex_imag)
+    {
+      return is_local_composite_access(to_complex_imag_expr(expr).op());
+    }
+    else
+    {
+      // matches ID_address_of, ID_dereference, etc.
+      return false;
+    }
+  }
+};
+
+// For goto_functions
+class function_cfg_infot : public cfg_infot
+{
+public:
+  function_cfg_infot(const goto_functiont &_goto_function)
+    : dirty_analysis(_goto_function), locals(_goto_function)
+  {
+    parameters.insert(
+      _goto_function.parameter_identifiers.begin(),
+      _goto_function.parameter_identifiers.end());
+  }
+
+  /// Returns true iff `ident` is a local or parameter of the goto_function.
+  bool is_local(const irep_idt &ident) const override
+  {
+    return locals.is_local(ident) ||
+           (parameters.find(ident) != parameters.end());
+  }
+
+  /// Returns true iff the given `ident` is either not a goto_function local
+  /// or is a local that is dirty.
+  bool is_not_local_or_dirty_local(const irep_idt &ident) const override
+  {
+    if(is_local(ident))
+      return dirty_analysis.get_dirty_ids().find(ident) !=
+             dirty_analysis.get_dirty_ids().end();
+    else
+      return true;
+  }
+
+private:
+  const dirtyt dirty_analysis;
+  const localst locals;
+  std::set<irep_idt> parameters;
```

Comment:
We have a hash function for ireps available, so we might as well just make these sets unordered:
```suggestion
  std::unordered_set<irep_idt> parameters;
```